### ISSUE TITLE
Ability to specify conflicting parameter names?

### ISSUE THREAD

**Issue by FISHMANPET** [pester/Pester#1290]
## 1. General summary of the issue

I believe this is related to #1007.  I see in the fix for that, the variable $script:ConflictingParameterNames was hardcoded which currently just contains PSEdition.  I happen to be mocking a function that contains a parameter called host, and I'm getting the same error that was originally reported in that bug.  As far as I can tell there's no way to specify conflicting parameters.  If manually add "host" to list of ConflictingParameterNames in Mock.ps1 my test succeeds.

This is the function I'm trying to mock:
```
function Send-SplunkHEC
	{
		<#
			.Synopsis
				Send event to Splunk HTTP Event Collector
			.DESCRIPTION
				Send event to Splunk HTTP Event Collector
			.PARAMETER uri
				URI for HEC endpoint
			.PARAMETER header
				contains auth token
			.PARAMETER host
				Part of Splunk Metadata for event.  Device data being sent from
			.PARAMETER source
				Part of Splunk Metadata for event.  Source 
			.PARAMETER sourceType
				Part of Splunk Metadata for event.  SourceType
			.PARAMETER metadata
				Part of Splunk Metadata for event.   Combination of host,source,sourcetype in performatted hashtable, will be comverted to JSON
			.PARAMETER eventData
				Event Data in hastable or pscustomeobject, will be comverted to JSON
		#>
		[CmdletBinding()]
		Param
		(
			# Param1 help description
			[Parameter(Mandatory)]
			[string]$uri,

			[Parameter(Mandatory)]
			[Collections.Hashtable]$header,

			[Parameter(Mandatory,ParameterSetName='Components')]
			[String]$host,

			[Parameter(Mandatory,ParameterSetName='Components')]
			[String]$source,

			[Parameter(Mandatory,ParameterSetName='Components')]
			[String]$sourceType,

			[Parameter(Mandatory,ParameterSetName='hashtable')]
			[Collections.Hashtable]$metadata,

			# This can be [Management.Automation.PSCustomObject] or [Collections.Hashtable]
			[Parameter(Mandatory)]
			$eventData
		)

		Begin{}
		Process
		{
			if ($metadata){$bodySplunk = $metadata}
			else {$bodySplunk = @{'host' = $host;'source' = $source;'sourcetype' = $sourcetype}}
			#Splunk takes time in Unix Epoch format, so first get the current date,
			#convert it to UTC (what Epoch is based on) then format it to seconds since January 1 1970.
			#Without converting it to UTC the date would be offset by a number of hours equal to your timezone's offset from UTC
			$bodySplunk['time'] = (Get-Date).toUniversalTime() | Get-Date -UFormat %s
			$bodySplunk['event'] = $eventData
			try {
				$response = Invoke-RestMethod -Uri $uri -Headers $header -UseBasicParsing -Body ($bodySplunk | ConvertTo-Json) -Method Post
			} catch {
				#$errordetail = Get-WebReqErrorDetails $_
				#$errordetail
				Write-Warning $_
			}
			if ($response.text -ne 'Success' -or $response.code -ne 0){throw "Failed to submit to Splunk HEC $($response)"}
			else{return $true}
		}
		End{}
	}
```
(from this module: https://github.com/umn-microsoft-automation/UMN-Common)

This is the function I'm trying to test:
```
function Send-ValidationError {
    param (
        $msg,
        $splunkParam,
        [Parameter(Mandatory=$false)]$data
    )
    $splunkParam.eventData["debugMsg"] = $msg
    if ($data) {
        $splunkParam.eventData["debugData"]
    }
    $null = Send-SplunkHEC @splunkParam
}
```

My mock and test is very simple:
```
Describe "Send-ValidationError" {
    It "should return nothing" {
        Mock Send-SplunkHEC {return $true}
        $splunkParam = @{
            uri = 'uri'
            metadata = @{metadata = 'metadata'}
            header = @{header = 'header'}
            eventData = @{jobID=(New-Guid).Guid}
        }
        Send-ValidationError -msg "test" -splunkParam $splunkParam | Should BeNullOrEmpty
    }
}
```
## 2. Describe Your Environment

Pester version     : 4.7.3 C:\Program Files\WindowsPowerShell\Modules\Pester\4.7.3\Pester.psd1
PowerShell version : 5.1.14393.2828
OS version         : Microsoft Windows NT 10.0.14393.0

## 3. Expected Behavior

When I add 'host' to the list of conflicting parameters in Mock.ps1, this is output I get
```
  Describing Send-ValidationError
    [+] should return nothing 29ms
Tests completed in 124ms
Tests Passed: 1, Failed: 0, Skipped: 0, Pending: 0, Inconclusive: 0
```

## 4.Current Behavior

When running 4.7.3 as shipped from the gallery this is the output:
```
  Describing Send-ValidationError
    [-] should return nothing 52ms
      at <ScriptBlock>, C:\<blah>.Tests.ps1: line 12
      12:         Send-ValidationError -msg "test" -splunkParam $splunkParam | Should BeNullOrEmpty
      
      Cannot overwrite variable host because it is read-only or constant.
Tests completed in 152ms
Tests Passed: 0, Failed: 1, Skipped: 0, Pending: 0, Inconclusive: 0
```

## 5. Possible Solution

Adding hosts to the list of conflicting parameter names fixed my problem, but I don't know enough about the code to say if adding some way to set that dynamically is the best way to solve this problem, especially since I see there's a lot of work currently happening around improving Mocks.

I also wrote my first Pester test this morning but if adding a way to set that dynamically is determined to be the best solution I can take a stab at it.

**Comment by renehernandez** [comment]
@FISHMANPET Thanks for bringing this up to our attention.

While working on #1007, we discussed the possibility of having the default `$script:ConflictingParameterName` to account for multiple variables, but we agreed that it could have ramifications that we were not able to account for at the time.

There are several things we could here:

1- Just add *Host* as a new entry into the `$script:ConflictingParameterName` array, but that won't help us with future occurrences with similar conflicting read-only/constants variables.
2- Coming up with a method for keeping in sync the [PowerShell automatic variables](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-6) in the `$script:ConflictingParameterName` array (since so far all the issues related to this stem from those variables), or at least the ones that would seem likely to be of conflict.
3- As you suggested, adding a method that could expand the default list in `$script:ConflictingParameterName` seems like a good idea,  but from the top of mind, we would need to take into account, at least, the scoping of this dynamic modifications, would they be locally to `It`, `Context` or `Describe` blocks? Or could that be specified as well?

I think that options 2 and 3 could be both implemented, which would give us a larger default coverage and the ability for end users to solve tackle this on their own.

@nohwnd Any thoughts?
@FISHMANPET If you want to give it a try to tackle this, give it a try and feel free to reach out for help if you needed

**Comment by nohwnd** [comment]
@renehernandez  All of those are valid options, if I had to choose I'd go with `1)`, just adding the common read-only/constant automatic variables in the list, the list is short and does not grow that much.

@FISHMANPET  Is it possible that you fix it on your side, the same way we do it? (you give the parameter a non-conflicting name, and alias it to `host`).

**Comment by FISHMANPET** [comment]
I guess I forgot to add "fixes #1290" into my original PR but since that's merged is there anything left to with this or can this be closed?

**Comment by renehernandez** [comment]
@FISHMANPET Good point. I guess I am used to issues being closed automatically that I missed it. Will close now
